### PR TITLE
[misc] Check code format using GitHub Actions/Workflows

### PR DESCRIPTION
**Pull request by archibate**
Related issue = #956

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

**Comment by archibate** [comment]
For now I have to rebuild Taichi to run `ti format`.

**Comment by archibate** [comment]
Now separated safely, in `misc/format.py`, just lose that we can't format by diff.

**Comment by archibate** [comment]
Will merging cause format break?

**Comment by archibate** [comment]
Hello? How do you make it work in misc/format_server.py?

**Comment by yuanming-hu** [comment]
> Hello? How do you make it work in misc/format_server.py?

I think I added an ssh-key to the machine hosting the format server. This is probably not applicable here since every time you get a new machine...

**Comment by archibate** [comment]
Maybe we can
1. Add a same ssh-key each time?
2. Use https, input secrets.passwd.

**Comment by rexwangcc** [comment]
For 
> 1. Add a same ssh-key each time?

Adding the following to the action script hack might help?

```
mkdir -p ~/.ssh
echo "${{ secrets.ssh_key }}" > ~/.ssh/ssh_key
chmod 600 ~/.ssh/ssh_key
echo """
Host github.com
	HostName github.com
	User git
	IdentityFile ~/.ssh/ssh_key
""" > ~/.ssh/config
```

**Comment by archibate** [comment]
after 44 useless tries I realized that actions are expected to be read-only? I.e. return true or false, instead of push.

**Comment by yuanming-hu** [comment]
> after 44 useless tries I realized that actions are expected to be read-only? I.e. return true or false, instead of push.

Is this post helpful to you? https://peterevans.dev/posts/github-actions-how-to-automate-code-formatting-in-pull-requests/

**Comment by rexwangcc** [comment]
Based on discussion in #621 , I thought the goal is to have both linter (return bool) and formatter (act similar to what the format server does) with Github Actions? (Also not sure if you notice, there's a ssh key got pushed in the commits, might want to remove and revoke it :octocat: )

**Comment by rexwangcc** [comment]
Did some quick testing in my personal testing repos, it seems https://peterevans.dev/posts/github-actions-how-to-automate-code-formatting-in-pull-requests/ won't work in taichi's fork-driven development process, since the token approaches does not support trigger actions from forks. 

A potential working solution I'm testing now is defining the SSH explicitly each time a workflow runs, see the latest commit of: https://github.com/rexwangcc/Joyful-Deep-Leaning-I/pull/1

It also successfully triggered PRs from forks, but I think the branch resolving with git needs to be further tweaked before it can work with push: https://github.com/rexwangcc/Joyful-Deep-Leaning-I/pull/3 

Generating SSH keys for the gardener will need help from maintainers tho. (Or we could simply re-use the one the format server is using)

**Comment by archibate** [comment]
I use check-only, no push-back attempt for now, and it works!

File: misc/format.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        print(Fore.YELLOW +
+              'Did not find any clang-format executable, skipping C++ files',
+              file=sys.stderr)
+    else:
+        print('C++ formatter: {}{}'.format(Fore.GREEN, result))
+    print(Style.RESET_ALL)
+    return result
+
+
+def main(all=False, diff=None):
```
Comment (k-ye) [inline misc/format.py:58]: BTW, can we also make `ti format` rely on this file's implementation, so that we don't have to repeat the code? (Not urgent)

Comment (archibate) [reply misc/format.py:58]: Will in a sep PR for all you want in #647 (also for `misc/build_doc.py` like this).

File: misc/release.sh [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+set -e
+
+git tag ${tag?no tag specified}
+git push --tags -u origin master
+git checkout stable
+git merge master
+git push -u origin stable
```
Comment (archibate) [inline misc/release.sh:8]:
```suggestion
git push -u origin stable
clear
python misc/make_changelog.py $tag
```

File: misc/release.sh [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+clear
+python misc/make_changelog.py $tag
```
Comment (yuanming-hu) [inline misc/release.sh:10]: I believe this part should probably in another PR...

File: misc/format.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+            formatted += l + '\n'
+    while len(formatted) and formatted[-1] == '\n':
+        formatted = formatted[:-1]
+    formatted += '\n'
+    with open(fn, 'w') as f:
+        f.write(formatted)
+
+
+def _find_clang_format_bin():
+    candidates = ['clang-format-6.0', 'clang-format']
```
Comment (yuanming-hu) [inline misc/format.py:37]: This is probably not an issue of this PR - should we stick to `clang-format-6.0` only? Different versions of `clang-format` can give you different results.

Comment (archibate) [reply misc/format.py:37]: The action server now seems to use clang-format-6.0.
Can we only use cf6? If user can't deal with cf6 installation, they should use our format server.
This req discussion and want to talk in a sep pr anyway.

File: misc/release.sh (hunk 3 above)
Comment (archibate) [reply misc/release.sh:10]: t4po, I guess you want this for a long time: github.com/actions/create-release, iapr of course.

**Comment by archibate** [comment]
merging now! hope people enjoy this check.

**Pull request merged**